lib/rpc/src/middleware.rs

```diff
@@ -0,0 +1,99 @@
+use std::time::{Duration, Instant};
+
+use crate::metrics::API_METRICS;
+use jsonrpsee::core::middleware::{Batch, Notification};
+use jsonrpsee::server::middleware::rpc::{RpcService, RpcServiceT};
+use jsonrpsee::types::Request;
+
+#[derive(Debug)]
+pub enum CallKind {
+    Call,
+    Notification,
+    Batch,
+}
+
+#[derive(Clone)]
+pub struct Monitoring {
+    inner: RpcService,
+}
+
+impl Monitoring {
+    pub fn new(inner: RpcService) -> Self {
+        Self { inner }
+    }
+}
+
+impl RpcServiceT for Monitoring {
+    type MethodResponse = <RpcService as RpcServiceT>::MethodResponse;
+    type NotificationResponse = <RpcService as RpcServiceT>::NotificationResponse;
+    type BatchResponse = <RpcService as RpcServiceT>::BatchResponse;
+
+    fn call<'a>(
+        &self,
+        request: Request<'a>,
+    ) -> impl Future<Output = Self::MethodResponse> + Send + 'a {
+        let method = request.method_name().to_owned();
+        let fut = self.inner.call(request);
+
+        async move {
+            let started = Instant::now();
+            let out = fut.await;
+            let output_size = out.as_json().get().len();
+
+            log_and_report(CallKind::Call, &method, started.elapsed(), output_size);
+            out
+        }
+    }
+
+    fn batch<'a>(
+        &self,
+        requests: Batch<'a>,
+    ) -> impl Future<Output = Self::BatchResponse> + Send + 'a {
+        let fut = self.inner.batch(requests);
+
+        async move {
+            let started = Instant::now();
+            let out = fut.await;
+            let output_size = out.as_json().get().len();
+
+            log_and_report(CallKind::Batch, "batch", started.elapsed(), output_size);
```

Comment:
very interesting! Is this called on a batched rpc request?? I was thinking RPC batch is just a collection of method invocations - so I'd expect them to be tracked separately. But here you have just one `fut` still? I suppose it iterates over the inner requests somehow inside?

This makes our tracking work poorly for batched requests - they won't even be counted towards the individual methods. Can we somehow "open" that batch here?

I'd also be interested in some batching-specific metrics like average/heatmap number of method invokcations per batch.

Or maybe I completely misunderstood what bathc is :)

Reply:
I am pretty sure you can achieve this if you re-implement `Monitoring::batch` by copying `RpcService::batch` implementation and adapting it to call `Monitoring::call` instead of `RpcService::call`